crates/kitsune_p2p/dht/src/region_set.rs

```diff
@@ -0,0 +1,313 @@
+//! "LTCS" is an acronym standing for "Logarithmic Time, Constant Space",
+//! and it refers to our current scheme of partitioning spacetime during gossip, which
+//! is to use a constant number of SpaceSegments (8..16) and a logarithmically
+//! growing number of TimeSegments, with larger segments to cover older times.
+//! In the future we may have other schemes.
+
+mod ltcs;
+
+pub use ltcs::*;
+
+use crate::{error::GossipResult, spacetime::*};
+
+use crate::region::{Region, RegionBounds, RegionCoords, RegionData, RegionDataConstraints};
+
+/// The generic definition of a set of Regions.
+/// The current representation is very specific to our current algorithm,
+/// but this is an enum to make room for a more generic representation, e.g.
+/// a simple Vec<Region>, if we want a more intricate algorithm later.
+#[derive(Debug, derive_more::From)]
+#[cfg_attr(feature = "test_utils", derive(Clone))]
+pub enum RegionSet<T: RegionDataConstraints = RegionData> {
+    /// Logarithmic Time, Constant Space.
+    Ltcs(RegionSetLtcs<T>),
+}
+
+impl<D: RegionDataConstraints> RegionSet<D> {
+    /// The number of regions specified
+    pub fn count(&self) -> usize {
+        match self {
+            Self::Ltcs(set) => set.count(),
+        }
+    }
+
+    /// can be used to pair the generated coords with stored data.
+    pub fn region_coords(&self) -> impl Iterator<Item = RegionCoords> + '_ {
+        match self {
+            Self::Ltcs(set) => set.coords.region_coords_flat().map(|(_, coords)| coords),
+        }
+    }
+
+    /// Iterator over all Regions
+    pub fn regions(&self) -> impl Iterator<Item = Region<D>> + '_ {
+        match self {
+            Self::Ltcs(set) => set.regions(),
+        }
+    }
+
+    /// The RegionSet can be used to answer questions about more regions than
+    /// just the ones specified: If a larger region is queried, and this set contains
+    /// a set of regions which over that larger region, then the larger region
+    /// can be dynamically constructed.
+    ///
+    /// This allows agents with differently computed RegionSets to still engage
+    /// in gossip without needing to recompute regions.
+    pub fn query(&self, _bounds: &RegionBounds) -> ! {
+        unimplemented!("only implement after trying naive database-only approach")
+    }
+
+    /// In order for this RegionSet to be queryable, new data needs to be
+    /// integrated into it to avoid needing to recompute it from the database
+    /// on each query.
+    pub fn update(&self, _c: SpacetimeQuantumCoords, _d: D) -> ! {
+        unimplemented!("only implement after trying naive database-only approach")
+    }
+
+    /// Find a set of Regions which represents the intersection of the two
+    /// input RegionSets.
+    pub fn diff(self, other: Self) -> GossipResult<Vec<Region<D>>> {
+        match (self, other) {
+            (Self::Ltcs(left), Self::Ltcs(right)) => left.diff(right),
```

Comment:
this match is gonna be *super* hairy if we ever get beyond two schemes... but we can cross that bridge if it ever happens : )

Reply:
I imagine one of the schemes will eventually be a totally arbitrary collection of regions, and once that lands, we can map any other scheme onto that one.